### PR TITLE
ARROW-3656: [C++] Allow whitespace in numeric CSV fields

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

File: cpp/src/arrow/csv/converter.cc [hunk 1]
```diff
@@ -44,6 +44,13 @@ Status GenericConversionError(const std::shared_ptr<DataType>& type, const uint8
   return Status::Invalid(ss.str());
 }
 
+inline bool IsWhitespace(uint8_t c) {
+  if (ARROW_PREDICT_TRUE(c > 0x20)) {
```
Comment (alendit) [inline cpp/src/arrow/csv/converter.cc]: Maybe `c > ' ` or even `c > ' ' && c > '\t'` here to make it more explicit? The last one should get optimized to a single comparison anyway.

File: cpp/src/arrow/csv/converter.cc [hunk 2]
```diff
@@ -275,10 +282,29 @@ Status NumericConverter<T>::Convert(const BlockParser& parser, int32_t col_index
   StringConverter<T> converter;
 
   auto visit = [&](const uint8_t* data, uint32_t size, bool quoted) -> Status {
+    // XXX should quoted values be allowed at all?
     value_type value;
     if (IsNull(data, size, quoted)) {
       return builder.AppendNull();
```
Comment (alendit) [inline cpp/src/arrow/csv/converter.cc]: Even though it's not a part of the changeset: why do we use the safe method here, but an unsafe one at around 312? Couldn't it both be unsafe?

Comment (pitrou) [reply cpp/src/arrow/csv/converter.cc]: Yes, good point.

**Comment by pitrou** [comment]
@wesm Any concern here? Otherwise I'll rebase and merge soon.

**Comment by pitrou** [comment]
+1, will merge.

**Pull request closed without merging**